File: apps/ios/Data/Sources/Data/DI/DataModule.swift
import Foundation
import Factory
import Supabase
import PowerSync
import Domain

// MARK: - DataModule (Factory)

public extension Container {
    /// Which backend this build talks to. See Config/SanvyaConfig.swift — the
    /// URL and key below used to be literals right here.
    var sanvyaConfig: Factory<SanvyaConfig> {
        self { BundleSanvyaConfig() }.singleton
    }

    var supabaseClient: Factory<SupabaseClient> {
        self {
            let config = self.sanvyaConfig()
            return SupabaseClient(
                supabaseURL: config.supabaseURL,
                supabaseKey: config.supabaseAnonKey
            )
        }.singleton
    }

    var powerSyncDatabase: Factory<PowerSyncDatabaseProtocol> {
        self {
            let powersyncTables = PocketCareSchema.tables.map { tableDef in
                Table(
                    name: tableDef.name,
                    columns: tableDef.columns.map { col in
                        let type: ColumnData
                        switch col.type {
                        case .text: type = .text
                        case .integer: type = .integer
                        case .real: type = .real
                        }
                        return Column(name: col.name, type: type)
                    },
                    indexes: tableDef.indexes.map { idx in
                        Index(
                            name: idx.name,
                            columns: idx.columns.map { idxCol in
                                IndexedColumn(column: idxCol.name, ascending: idxCol.ascending)
                            }
                        )
                    },
                    localOnly: tableDef.localOnly,
                    insertOnly: tableDef.insertOnly
                )
            }
            let schema = Schema(tables: powersyncTables)
            return try! PowerSyncDatabase(schema: schema, dbFilename: "pocketcare.db")
        }.singleton
    }

    var supabaseConnector: Factory<SupabaseConnector> {
        self {
            SupabaseConnector(
                client: self.supabaseClient(),
                powerSyncUrl: self.sanvyaConfig().powerSyncURL
            )
        }.singleton
    }

    var authRepository: Factory<AuthRepository> {
        self {
            AuthRepositoryImpl(
                client: self.supabaseClient(),
                config: self.sanvyaConfig()
            )
        }.singleton
    }

    var ledgerRepository: Factory<LedgerRepository> {
        self { LedgerRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var budgetRepository: Factory<BudgetRepository> {
        self { BudgetRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var creditCardRepository: Factory<CreditCardRepository> {
        self { CreditCardRepository(db: self.powerSyncDatabase(), transactions: self.ledgerRepository()) }.singleton
    }

    var splitsRepository: Factory<SplitsRepository> {
        self { SplitsRepository(db: self.powerSyncDatabase(), ledger: self.ledgerRepository()) }.singleton
    }

    // The two catch-up engines. Both take repositories rather than reaching for
    // the database directly for the writes -- createTransaction() carries the
    // overdraft guard and the transfer/items validation, and an engine that
    // bypassed it would post rows the app itself would refuse.
    var recurringRepository: Factory<RecurringRepository> {
        self {
            RecurringRepository(
                db: self.powerSyncDatabase(),
                ledger: self.ledgerRepository(),
                splits: self.splitsRepository()
            )
        }.singleton
    }

    var loanAutoPostRepository: Factory<LoanAutoPostRepository> {
        self {
            LoanAutoPostRepository(db: self.powerSyncDatabase(), ledger: self.ledgerRepository())
        }.singleton
    }

    // Takes the database (the disclosure audit trail is synced and read from
    // local SQLite) and UserDefaults (the masked hint is a device pref, exactly
    // as web keeps it in localStorage) alongside the client the Edge Function
    // calls go through.
    var upiRepository: Factory<UpiRepository> {
        self {
            let auth = self.authRepository()
            return UpiRepository(
                client: self.supabaseClient(),
                db: self.powerSyncDatabase(),
                getUserId: { auth.currentUserId }
            )
        }.singleton
    }

    var invitesRepository: Factory<InvitesRepository> {
        self { InvitesRepository(client: self.supabaseClient()) }.singleton
    }

    var assistantRepository: Factory<AssistantRepository> {
        self {
            AssistantRepository(
                db: self.powerSyncDatabase(),
                client: self.supabaseClient(),
                ledgerRepository: self.ledgerRepository(),
                goalsRepository: self.goalsRepository(),
                budgetRepository: self.budgetRepository(),
                subscriptionsRepository: self.subscriptionsRepository(),
                splitsRepository: self.splitsRepository()
            )
        }.singleton
    }

    var goalsRepository: Factory<GoalsRepository> {
        self { GoalsRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var investmentsRepository: Factory<InvestmentsRepository> {
        self { InvestmentsRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var loansRepository: Factory<LoansRepository> {
        self { LoansRepository(db: self.powerSyncDatabase()) }.singleton
    }

    /// The shell's bell badge and the notifications inbox.
    var notificationsRepository: Factory<NotificationsRepository> {
        self { NotificationsRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var prefsRepository: Factory<PrefsRepository> {
        self { PrefsRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var repairRepository: Factory<RepairRepository> {
        self {
            let auth = self.authRepository()
            return RepairRepository(
                db: self.powerSyncDatabase(),
                client: self.supabaseClient(),
                getUserId: { auth.currentUserId ?? "" }
            )
        }.singleton
    }

    // Zero-trust field encryption: the key lifecycle, the support grants, and
    // the Keychain that holds an unlock across app termination. The support
    // public key is passed in rather than injected so the repository never
    // learns what SanvyaConfig is.
    var secureKeyStore: Factory<SecureKeyStore> {
        self { SecureKeyStore() }.singleton
    }

    var securityRepository: Factory<SecurityRepository> {
        self {
            let auth = self.authRepository()
            return SecurityRepository(
                db: self.powerSyncDatabase(),
                client: self.supabaseClient(),
                store: self.secureKeyStore(),
                getUserId: { auth.currentUserId },
                supportPublicJwk: self.sanvyaConfig().supportPublicJwk
            )
        }.singleton
    }

    var pushRepository: Factory<PushRepository> {
        self { SupabasePushRepository(client: self.supabaseClient()) }.singleton
    }

    var subscriptionsRepository: Factory<SubscriptionsRepository> {
        self { SubscriptionsRepository(db: self.powerSyncDatabase()) }.singleton
    }

    var feedbackRepository: Factory<FeedbackRepository> {
        self { FeedbackRepository(db: self.powerSyncDatabase()) }.singleton
    }

    // Task #62 (Receipt Scan capture) -- ReceiptsRepository itself already
    // existed (P2.5) but was never registered here, so nothing could ever
    // resolve it via DI. That's why the old ReceiptScanView.swift was a
    // hardcoded fixture with no repository wiring at all.
    var receiptsRepository: Factory<ReceiptsRepository> {
        self {
            let auth = self.authRepository()
            return ReceiptsRepository(
                db: self.powerSyncDatabase(),
                getUserId: { auth.currentUserId ?? "" },
                client: self.supabaseClient()
            )
        }.singleton
    }
}

File: apps/android/data/src/main/kotlin/com/sanvya/app/data/di/DataModule.kt
package com.sanvya.app.data.di

import android.content.Context
import com.sanvya.app.data.auth.AuthRepository
import com.sanvya.app.data.auth.AuthRepositoryImpl
import com.sanvya.app.data.config.SanvyaConfig
import com.sanvya.app.data.config.sanvyaConfig
import com.powersync.PowerSyncDatabase
import com.sanvya.app.data.db.DatabaseManager
import com.sanvya.app.data.sync.SupabaseConnector
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.auth.Auth
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.functions.Functions
import com.sanvya.app.data.repository.LedgerRepository
import com.sanvya.app.data.repository.BudgetRepository
import com.sanvya.app.data.repository.CreditCardRepository
import com.sanvya.app.data.repository.GoalsRepository
import com.sanvya.app.data.repository.InvestmentsRepository
import com.sanvya.app.data.repository.LoansRepository
import com.sanvya.app.data.repository.ReceiptsRepository
import com.sanvya.app.data.repository.RepairRepository
import com.sanvya.app.data.repository.LoanAutoPostRepository
import com.sanvya.app.data.repository.RecurringRepository
import com.sanvya.app.data.repository.SplitsRepository
import com.sanvya.app.data.repository.SubscriptionsRepository
import org.koin.android.ext.koin.androidContext
import org.koin.dsl.module

val dataModule = module {
    // Which backend this build talks to. See config/SanvyaConfig.kt -- the
    // URLs and keys below used to be literals right here.
    single<SanvyaConfig> { sanvyaConfig() }

    single<AuthRepository> { AuthRepositoryImpl(get()) }

    single<SupabaseClient> {
        val config: SanvyaConfig = get()
        createSupabaseClient(
            supabaseUrl = config.supabaseUrl,
            supabaseKey = config.supabaseAnonKey,
        ) {
            install(Auth) {
                // The custom-scheme callback the OAuth provider returns to.
                // supabase-kt builds `<scheme>://<host>` from these and hands
                // it to the provider; MainActivity forwards the resulting
                // Intent back in via handleDeeplinks(). Both halves read the
                // same config, and the manifest intent filter is generated
                // from the same two Gradle properties, so the three cannot
                // drift out of agreement.
                scheme = config.authRedirectScheme
                host = config.authRedirectHost
            }
            install(Postgrest)
            install(Functions)
        }
    }

    single<PowerSyncDatabase> {
        DatabaseManager.createDatabase(androidContext())
    }

    single<SupabaseConnector> {
        SupabaseConnector(
            client = get(),
            powerSyncUrl = get<SanvyaConfig>().powerSyncUrl,
        )
    }

    single { LedgerRepository(get()) }
    single { BudgetRepository(get()) }
    single { CreditCardRepository(get(), get()) }
    single { GoalsRepository(get()) }
    single { InvestmentsRepository(get()) }
    single { LoansRepository(get()) }
    single { SplitsRepository(get(), get()) }
    // The two catch-up engines. Both take repositories rather than reaching for
    // the database directly for the writes -- createTransaction() carries the
    // overdraft guard and the transfer/items validation, and an engine that
    // bypassed it would post rows the app itself would refuse.
    single { RecurringRepository(db = get(), ledger = get(), splits = get()) }
    single { LoanAutoPostRepository(db = get(), ledger = get()) }
    single { com.sanvya.app.data.repository.UpiRepository(get()) }
    single { com.sanvya.app.data.repository.InvitesRepository(get()) }
    single { com.sanvya.app.data.repository.FeedbackRepository(get()) }
    single {
        com.sanvya.app.data.repository.AssistantRepository(
            db = get(),
            client = get(),
            ledgerRepository = get(),
            goalsRepository = get(),
            budgetRepository = get(),
            subscriptionsRepository = get(),
            splitsRepository = get(),
        )
    }
    single { SubscriptionsRepository(get()) }
    single { com.sanvya.app.data.repository.PrefsRepository(get()) }
    // Settings' own data access — keeps SupabaseClient/PowerSyncDatabase out of :app.
    single { com.sanvya.app.data.repository.SettingsRepository(get(), get()) }
    // The dashboard's "Worth a look" strip -- one row of counts, one query.
    single { com.sanvya.app.data.repository.SuggestionsRepository(get()) }
    // The app's ONE sync status: online + connected + hasSynced + lastSyncedAt,
    // on a single poll loop. Replaces four independent 400 ms pollers and the
    // shell's private connectivity callback -- see SyncStatusRepository.kt.
    single { com.sanvya.app.data.sync.SyncStatusRepository(androidContext(), get()) }
    // The thing that actually connects PowerSync to the server. Until this
    // existed, `SupabaseConnector` was built here and handed to nothing --
    // see SyncBootstrap.kt.
    single { com.sanvya.app.data.sync.SyncBootstrap(get(), get(), get()) }
    // The shell's bell badge and the notifications inbox.
    single { com.sanvya.app.data.repository.NotificationsRepository(get()) }
    single {
        val auth: AuthRepository = get()
        ReceiptsRepository(db = get(), getUserId = { auth.currentUserId.value ?: "" }, client = get())
    }
    single<com.sanvya.app.domain.repository.PushRepository> { com.sanvya.app.data.repository.SupabasePushRepository(get()) }
    single {
        val auth: AuthRepository = get()
        RepairRepository(db = get(), client = get(), getUserId = { auth.currentUserId.value ?: "" })
    }
}
